astronify/simulator/add_flare_signal.py

```diff
@@ -5,12 +5,119 @@
 .. moduleauthor:: Scott W. Fleming <fleming@stsci.edu>
 """
 
-def add_flare_signal(fluxes):
+import math
+import numpy as np
+
+def add_flare_signal(fluxes, flare_time, flare_amp, flare_halfwidth):
     """
+    Model is based on Davenport et al.
+    (https://ui.adsabs.harvard.edu/abs/2014ApJ...797..122D/abstract)
+
+    F_rise = 1 + 1.941(t) - 0.175(t^2) - 2.246(t^3) - 1.125(t^4)
+    where t = time at half the maximum flux, after normalizing such that
+    the flux at t=0 is the peak of the flare.
+
+    F_decay = 0.6890*e^(-1.600*t) + 0.3030*e^(-0.2783*t)
+    where t = time at half the maximum flux, after normalizing such that
+    the flux at t=0 is the peak of the flare.
+
     :param fluxes: Array of fluxes to add the flare signal to.
     :type fluxes: numpy.ndarray
 
+    :param flare_time: Time corresponding to the maximum flare flux.
+    :type flare_time: int
+
+    :param flare_amp: The peak (maximum flux) of the flare.
+    :type flare_amp: float
+
+    :param flare_halfwidth: The flare half-width (measured in indices) that
+    corresponds to "t_1/2" in the Davenport et al. flare template.
+    :type flare_halfwidth: index
+
     :returns: numpy.ndarray -- The fluxes with the flare signal added.
     """
 
-    return fluxes
+    # This array will contain the fluxes of the flare to add to the input.
+    fluxes_to_add = np.zeros(fluxes.shape[0])
+
+    # We must be careful if the flare_halfwidth pushes us past the zero'th
+    # index though.  If it does, we temporarily prepend some extra indices.
+    truncated_start = flare_time - flare_halfwidth < 0
+    if truncated_start:
+        n_to_add_rise = abs(flare_time - flare_halfwidth + 1)
+        fluxes_to_add = np.concatenate([np.zeros(n_to_add_rise), fluxes_to_add])
+        # Need to move the 'flare_time' index over now...
+        flare_time += n_to_add_rise
+
+    # Where "n" = 6 in Davenport et al., where the decay phase is defined from
+    # t_1/2 = [0,6], but we may choose to extend further so that the decay gets
+    # as close to zero as possible.
+    n_t12 = 10
+
+    # Need to check if our decay tail will extend past the input array.
```

Comment:
Is there any reason not to just always have the decay extend exactly to the end of the input array? Then you don't have to worry about if it's getting sufficiently close to zero at the cutoff.

In any case, for larger flares you probably want n_t12 more than 10, for my flare injection code I used n_t12=100, although I doubt you would need to go that high for this application.

Reply:
Oops I forgot to include the new file, it should be in the repo now.

Reply:
And feel free to extend the variable to a larger default for now.  I would want it to extend to the end of the array, but I couldn't do the mental math to make that happen the way I had written it, so I took the easy way out and shrugged.